### ISSUE TITLE
Nullable int in Settings throws (and swallows) exception if empty

### ISSUE THREAD

**Issue by mcliment** [nopSolutions/nopCommerce#1902]
Investigating an exception that was thrown by our installation on almost every request, the problem was that `DateOfBirthMinimumAge` was null -kind of-.
This null values are stored in the settings table as empty values (as they're not nullable) and when retrieved, the `SettingsService` tries to convert the value using `GetNopCustomTypeConverter` that returns the default type converter for `Nullable<int>`. But this converter can't handle empty strings and internally throws an IndexOutOfRangeException -which is swallowed-.
This behavior is not desirable but I want to discuss the best option here:
* Don't allow nullable ints on settings
* Create a custom type converter for Nullable<int> that checks that the input value is not string empty before attempting the conversion
Any opinions on that?

**Comment by rchamorro** [comment]
Nullable ints are good expressing that something has not been set, better than setting 0, for example, so I think a custom type converter is the way to go.

**Comment by mcliment** [comment]
A third option may be to allow nulls in the Value of the Settings table...

**Comment by AndreiMaz** [comment]
@mcliment  Could you please clarify when and where exactly the exception is thrown?

**Comment by mcliment** [comment]
I found the exception using Stackify prefix. The exception thrown is this
```
System.IndexOutOfRangeException: Index was outside the bounds of the array. ---> System.Exception:  is not a valid value for Int32.
  at System.ComponentModel.BaseNumberConverter.ConvertFrom
--- End of inner exception stack trace ---
  at System.ComponentModel.BaseNumberConverter.ConvertFrom
  at System.ComponentModel.TypeConverter.IsValid
```

The explanation is that when you try to check if a value is valid (here [SettingsService#L373](https://github.com/nopSolutions/nopCommerce/blob/develop/src/Libraries/Nop.Services/Configuration/SettingService.cs#L373)), if it's a numeric value (nullable in this case) and the value to convert is an empty string, internally the method tries to make a conversion and throws an exception.

It's not a big issue but generally is not advisable to swallow exceptions (in this case in every request) and just a check that if the setting type it's a nullable value type and the setting value is and empty string the value will be null.

To see the exception in debug mode you need to disable 'Just my code' because it's thrown by .NET framework.

**Comment by skoshelev** [comment]
Fixed by [this commit](https://github.com/nopSolutions/nopCommerce/commit/abdc4a6d99e77900c8080f645739e045fd311ae9).
Closed #1902
